File: gravital-core/web/src/types/alert.ts
// 告警类型定义
export interface AlertRule {
  id: string
  rule_name: string
  description?: string
  severity: 'critical' | 'warning' | 'info'
  condition: string
  duration: number
  filters: Record<string, any>
  notification_config: Record<string, any>
  enabled: boolean
  created_at: string
  updated_at: string
}

// 告警事件
export interface AlertEvent {
  id: string
  rule_id: string
  rule_name: string
  severity: 'critical' | 'warning' | 'info'
  device_id: string
  device_name: string
  metric_name: string
  current_value: number
  threshold: number
  message: string
  status: 'firing' | 'resolved' | 'acknowledged'
  triggered_at: string
  resolved_at?: string
  acknowledged_at?: string
  acknowledged_by?: string
}

// 告警规则查询参数
export interface AlertRuleQuery {
  page?: number
  size?: number
  keyword?: string
  severity?: string
  enabled?: boolean
}

// 告警事件查询参数
export interface AlertEventQuery {
  page?: number
  size?: number
  severity?: string
  status?: string
  start_time?: string
  end_time?: string
}

// 通知渠道配置
export interface NotificationChannelConfig {
  channel: 'email' | 'webhook' | 'dingtalk' | 'wechat' | 'sms'
  enabled: boolean
  recipients: string[]
  template?: string
  config?: Record<string, any>
}

// 通知配置
export interface NotificationConfig {
  enabled: boolean
  channels: NotificationChannelConfig[]
  dedupe_interval?: number      // 去重间隔（秒）
  escalation_enabled?: boolean  // 是否启用升级
  escalation_after?: number     // 升级时间（秒）
  escalation_channels?: string[] // 升级通知渠道
}

// 创建告警规则请求
export interface AlertRuleCreate {
  rule_name: string
  description?: string
  severity: 'critical' | 'warning' | 'info'
  condition: string
  duration: number
  filters?: Record<string, any>
  notification_config?: NotificationConfig
  enabled?: boolean
}

// 告警聚合信息
export interface AlertAggregation {
  rule_id: number
  rule_name: string
  severity: 'critical' | 'warning' | 'info'
  description: string
  total_count: number
  firing_count: number
  acked_count: number
  first_fired: string
  last_fired: string
  devices: Array<{
    device_id: string
    device_name: string
    status: string
    triggered_at: string
  }>
}

File: gravital-core/internal/model/alert.go
package model

import "time"

// AlertRule 告警规则
type AlertRule struct {
	ID                 uint      `gorm:"primaryKey" json:"id"`
	RuleName           string    `gorm:"size:255;not null" json:"rule_name"`
	Enabled            bool      `gorm:"default:true" json:"enabled"`
	Severity           string    `gorm:"size:32" json:"severity"`
	Condition          string    `gorm:"type:text;not null" json:"condition"`
	Filters            JSONB     `gorm:"type:jsonb" json:"filters"`
	Duration           int       `json:"duration"`
	NotificationConfig JSONB     `gorm:"type:jsonb" json:"notification_config"`
	InhibitRules       JSONB     `gorm:"type:jsonb" json:"inhibit_rules"`
	MutePeriods        JSONB     `gorm:"type:jsonb" json:"mute_periods"`
	Description        string    `gorm:"type:text" json:"description"`
	CreatedBy          *uint     `json:"created_by"`
	CreatedAt          time.Time `json:"created_at"`
	UpdatedAt          time.Time `json:"updated_at"`
}

// AlertEvent 告警事件
type AlertEvent struct {
	ID               uint       `gorm:"primaryKey" json:"id"`
	AlertID          string     `gorm:"uniqueIndex;size:64;not null" json:"alert_id"`
	RuleID           uint       `gorm:"index" json:"rule_id"`
	Rule             *AlertRule `gorm:"foreignKey:RuleID" json:"rule,omitempty"`
	DeviceID         string     `gorm:"size:64;index" json:"device_id"`
	MetricName       string     `gorm:"size:255" json:"metric_name"`
	Severity         string     `gorm:"size:32" json:"severity"`
	Message          string     `gorm:"type:text" json:"message"`
	Labels           JSONB      `gorm:"type:jsonb" json:"labels"`
	TriggeredAt      time.Time  `gorm:"index" json:"triggered_at"`
	ResolvedAt       *time.Time `json:"resolved_at"`
	Status           string     `gorm:"size:32;index" json:"status"`
	NotificationSent bool       `gorm:"default:false" json:"notification_sent"`
	Acknowledged     bool       `gorm:"default:false" json:"acknowledged"`
	AcknowledgedBy   *uint      `json:"acknowledged_by"`
	AcknowledgedAt   *time.Time `json:"acknowledged_at"`
	Comment          string     `gorm:"type:text" json:"comment"`
	CreatedAt        time.Time  `json:"created_at"`
}

// AlertNotification 告警通知记录
type AlertNotification struct {
	ID           uint       `gorm:"primaryKey" json:"id"`
	AlertEventID uint       `gorm:"index" json:"alert_event_id"`
	Channel      string     `gorm:"size:64" json:"channel"`
	Recipient    string     `gorm:"size:255" json:"recipient"`
	Status       string     `gorm:"size:32" json:"status"`
	SentAt       *time.Time `json:"sent_at"`
	ErrorMessage string     `gorm:"type:text" json:"error_message"`
	CreatedAt    time.Time  `json:"created_at"`
}

// TableName 指定表名
func (AlertRule) TableName() string {
	return "alert_rules"
}

func (AlertEvent) TableName() string {
	return "alert_events"
}

func (AlertNotification) TableName() string {
	return "alert_notifications"
}

